src/CodeAnalysis/CSharp/MetaAnalyzers/CSharpImmutableObjectMethodAnalyzer.cs

```diff
@@ -0,0 +1,105 @@
+
+        public static DiagnosticDescriptor DoNotIgnoreReturnValueDiagnosticRule = new DiagnosticDescriptor(
+            DiagnosticIds.DoNotIgnoreReturnValueOnImmutableObjectMethodInvocation,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.AnalyzerCorrectness,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            description: s_localizableDescription,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(DoNotIgnoreReturnValueDiagnosticRule);
+
+        private static readonly string s_solutionFullName = @"Microsoft.CodeAnalysis.Solution";
+        private static readonly string s_projectFullName = @"Microsoft.CodeAnalysis.Project";
+        private static readonly string s_documentFullName = @"Microsoft.CodeAnalysis.Document";
+        private static readonly string s_syntaxNodeFullName = @"Microsoft.CodeAnalysis.SyntaxNode";
+        private static readonly string s_compilationFullName = @"Microsoft.CodeAnalysis.Compilation";
+
+        // A list of known immutable object names
+        private static readonly ImmutableArray<string> s_immutableObjectNames = ImmutableArray.Create(
+            s_solutionFullName,
+            s_projectFullName,
+            s_documentFullName,
+            s_syntaxNodeFullName,
+            s_compilationFullName);
+
+        private static readonly string s_Add = "Add";
+        private static readonly string s_Remove = "Remove";
+        private static readonly string s_Replace = "Replace";
+        private static readonly string s_With = "With";
+
+        private static readonly ImmutableArray<string> s_immutableMethodNames = ImmutableArray.Create(
+            s_Add, 
+            s_Remove, 
+            s_Replace, 
+            s_With);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterSyntaxNodeAction(AnalyzeInvocationForIgnoredReturnValue, SyntaxKind.InvocationExpression);
+        }
+
+        public void AnalyzeInvocationForIgnoredReturnValue(SyntaxNodeAnalysisContext context)
+        {
+            var model = context.SemanticModel;
+            var candidateInvocation = (InvocationExpressionSyntax)context.Node;
+
+            //We're looking for invocations that are direct children of expression statements
+            if (!(candidateInvocation.Parent is ExpressionStatementSyntax))
+            {
+                return;
+            }
+
+            //If we can't find the method symbol, quit
+            var methodSymbol = model.GetSymbolInfo(candidateInvocation).Symbol as IMethodSymbol;
```

Comment:
This is going to be expensive - you will probably end up invoking GetSymbolInfo on every invocation. I would recommend that you add a syntactic check before this to see if the node's identifier matches one of `s_immutableMethodNames`.